### PR TITLE
Fix: clean containers after `keadm reset`

### LINKED ISSUE #1973: Docker container not killed after `keadm reset`
**What happened**:

Hi!

In sometimes, I lost the connection between cloud and edge (restart my virtual machines). So I have to rejoin to the cloud by `keadm join`, after running `keadm reset`.

After several times, I found there are some old containers still running, even if I just executed `keadm reset`:

docker ps

**What you expected to happen**:

After run `keadm reset`, no related Docker container is running.

**How to reproduce it (as minimally and precisely as possible)**:

When deploy successfully, run `keadm reset` directlly, and run `docker ps`. You can see container(s) named `pause` in the list.

I also got almost the same result in cloud node. Maybe they have the same reason.

**Anything else we need to know?**:

By reading the source code, I found something related to that:

https://github.com/kubeedge/kubeedge/blob/ec23878a135bbad396ba847cc79cc9516663d712/keadm/cmd/keadm/app/cmd/reset.go#L76-L83

I want to ask **is there any plan still in progress at here?** I think maybe we can do something to avoid "zombie container", by:

```go
RunE: func(cmd *cobra.Command, args []string) error {
	// Tear down cloud node. It includes 
 	// 1. killing cloudcore process

 	// Tear down edge node. It includes 
 	// 1. killing edgecore process, but don't delete node from K8s 
	if err := TearDownKubeEdge(IsEdgeNode, reset.Kubeconfig); err != nil {
		return err
	}


	// For both cloud node and edge node:
	// 2. doing something else to cleanup
	if err2 := someCleaning(); err2 != nil {
		return err2
	}

	return nil
},
```

and

```go
// someCleaning does cleaning including kill old docker containers
func someCleaning() error {
	// TODO
	return nil
}
```

**Environment**:

Same environment for both cloud and edge nodes.

- KubeEdge version:  

```
KubeEdge v1.3.1
```

- Hardware configuration (e.g. `lscpu`):

```
[root@nginx-master-01 ~]# lscpu
Architecture:          	x86_64
CPU op-mode(s):        	32-bit, 64-bit
Byte Order:            	Little Endian
CPU(s):                	2
On-line CPU(s) list:   	0,1
Thread(s) per core:    	1
Core(s) per socket:    	1
座：                 	2
NUMA 节点：         	1
厂商 ID：           	GenuineIntel
CPU 系列：          	6
型号：              	94
型号名称：        	Intel(R) Core(TM) i5-6300HQ CPU @ 2.30GHz
步进：              	3
CPU MHz：             	2304.000
BogoMIPS：            	4608.00
超管理器厂商：  	VMware
虚拟化类型：     	完全
L1d 缓存：          	32K
L1i 缓存：          	32K
L2 缓存：           	256K
L3 缓存：           	6144K
NUMA 节点0 CPU：    	0,1
Flags:                 	fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon nopl xtopology tsc_reliable nonstop_tsc eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 avx2 smep bmi2 invpcid rdseed adx smap clflushopt xsaveopt xsavec arat spec_ctrl intel_stibp flush_l1d arch_capabilities
```

- OS (e.g. `cat /etc/os-release`):

```
[root@nginx-master-01 ~]# cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

- Kernel (e.g. `uname -a`):

```
[root@nginx-master-01 ~]# uname -a
Linux nginx-master-01 3.10.0-1062.1.2.el7.x86_64 #1 SMP Mon Sep 30 14:19:46 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

- Go version (e.g. `go version`): 
```
[root@nginx-master-01 ~]# go version
go version go1.14.6 linux/amd64
```

### PR DESCRIPTION
**Pull request by ESWZY**
**What type of PR is this?**

/kind bug
/kind design

**What this PR does / why we need it**:

We need to recycle staled containers to avoid resource occupation, which maybe cause memory leak or something.

You can see the details in the following issue.

**Which issue(s) this PR fixes**:

Fixes #1973 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: keadm/cmd/keadm/app/cmd/reset.go [hunk 1]
```diff
@@ -74,12 +76,21 @@ func NewKubeEdgeReset(out io.Writer, reset *types.ResetOptions) *cobra.Command {
...
-			// 1. killing cloudcore process
+			// 1. kill cloudcore/edgecore process.
+			// For edgecore, don't delete node from K8S
+			if err := TearDownKubeEdge(IsEdgeNode, reset.Kubeconfig); err != nil {
+				return err
+			}
+
+			// 2. remove containers managed by KE
+			_, _ = fmt.Fprintln(out, "[Test] Test here!")
+			if err2 := RemoveContainers(utilsexec.New()); err2 != nil {
```
Comment (GsssC) [inline keadm/cmd/keadm/app/cmd/reset.go:87]: should not delete containers when using `keadm reset` in cloud.

Comment (ESWZY) [reply keadm/cmd/keadm/app/cmd/reset.go:87]: Thanks for your correction! Is this correct?

```go
if IsEdgeNode {
	if err2 := RemoveContainers(utilsexec.New()); err2 != nil {
		return err2
	}
}
```

Comment (GsssC) [reply keadm/cmd/keadm/app/cmd/reset.go:87]: yes :), need to judge if on edgenode

**Comment by GsssC** [comment]
Consider whether the following items also needs to be deleted:
- /var/lib/edged
- /etc/kubeedge
- mount information

according to https://mritd.me/2016/11/21/kubeadm-other-problems/, this link is too old, maybe you also need to track what `kubeadm reset` do.

**Comment by GsssC** [comment]
: ), could u contact me with **DingTalk** once?@ESWZY

**Comment by ESWZY** [comment]
OK~ @GsssC

**Comment by GsssC** [comment]
@ESWZY : ) 
Stateful data at cloud:
- /etc/kubeedge
    - **undo**: refer to 'kubeadm', print tip to let user delete it manually.
- bin 
    - 'keadm reset' **already** deletes it
- crds in etcd 
    - get by 'kubectl get crds | grep kubeedge'
    - at master branch, 'keadm reset' **already** deletes these data.
- secret in etcd
    - get by 'kubectl get secret', at master branch, all can be delete by 'keadm reset'  now; 
    -  **undo**: Save secrets to /etc/kubeedge/secrets before deleting it.

Stateful data at edge:
- /etc/kubeedge
    - **undo**: refer to 'kubeadm', print tip to let user delete it manually.
- bin
    - 'keadm reset' **already** deletes it
- /var/lib/kubeedge
    - **undo**: print tip to get user admission, then delete it.
- /var/lib/edged
    - **undo**: refer to `kubeadm`
- mount information.
    - **undo**: refer to `kubeadm`

**Comment by GsssC** [comment]
Change `Enhancement Status` to `implemented` if you finish this improvment, : )
https://docs.google.com/spreadsheets/d/17cCqHE06Z2AG7AA_3UkQMOoLjK8v9vBUgaLsJoK7nu0/edit#gid=1175424024

**Comment by GsssC** [comment]
Change `Enhancement Status` to `implemented` if you finish this improvment, : )
https://docs.google.com/spreadsheets/d/17cCqHE06Z2AG7AA_3UkQMOoLjK8v9vBUgaLsJoK7nu0/edit#gid=1175424024

File: keadm/cmd/keadm/app/cmd/reset.go [hunk 2]
```diff
@@ -103,6 +118,48 @@ func TearDownKubeEdge(isEdgeNode bool, kubeConfig string) error {
...
+func RemoveContainers(isEdgeNode bool, execer utilsexec.Interface) error {
+	if !isEdgeNode {
+		return nil
+	}
+
+	criSocketPath, err := utilruntime.DetectCRISocket()
+	if err != nil {
+		return err
+	}
+	fmt.Println("[Test] Got criSocketPath", criSocketPath)
```
Comment (fisherxu) [inline keadm/cmd/keadm/app/cmd/reset.go:131]: Remove this line

File: keadm/cmd/keadm/app/cmd/reset.go [hunk 3]
```diff
@@ -103,6 +118,48 @@ func TearDownKubeEdge(isEdgeNode bool, kubeConfig string) error {
...
+
+	containerRuntime, err := utilruntime.NewContainerRuntime(execer, criSocketPath)
+	if err != nil {
+		return err
+	}
+	fmt.Println("[Test] Got containerRuntime", containerRuntime)
```
Comment (fisherxu) [inline keadm/cmd/keadm/app/cmd/reset.go:137]: Remove this line

File: keadm/cmd/keadm/app/cmd/reset.go [hunk 4]
```diff
@@ -103,6 +118,48 @@ func TearDownKubeEdge(isEdgeNode bool, kubeConfig string) error {
...
+
+	containers, err := containerRuntime.ListKubeContainers()
+	if err != nil {
+		return err
+	}
+	fmt.Println("[Test] Got containers", containers)
```
Comment (fisherxu) [inline keadm/cmd/keadm/app/cmd/reset.go:143]: ditto

File: keadm/cmd/keadm/app/cmd/reset.go [hunk 5]
```diff
@@ -74,12 +76,25 @@ func NewKubeEdgeReset(out io.Writer, reset *types.ResetOptions) *cobra.Command {
...
 		RunE: func(cmd *cobra.Command, args []string) error {
-			// Tear down cloud node. It includes
-			// 1. killing cloudcore process
+			// 1. kill cloudcore/edgecore process.
+			// For edgecore, don't delete node from K8S
+			if err := TearDownKubeEdge(IsEdgeNode, reset.Kubeconfig); err != nil {
+				return err
+			}
+
+			// 2. Remove containers managed by KE. Only for edge node.
```
Comment (fisherxu) [inline keadm/cmd/keadm/app/cmd/reset.go:85]: better to use `KubeEdge`

File: keadm/cmd/keadm/app/cmd/reset.go [hunk 6]
```diff
@@ -18,13 +18,15 @@ package cmd
...
-
 	"github.com/spf13/cobra"
+	"io"
 
 	"github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/common"
 	types "github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/common"
 	"github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/util"
+	phases "k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/reset"
+	utilruntime "k8s.io/kubernetes/cmd/kubeadm/app/util/runtime"
+	utilsexec "k8s.io/utils/exec"
```
Comment (fisherxu) [inline keadm/cmd/keadm/app/cmd/reset.go:29]: Please sort the imports

**Comment by ESWZY** [comment]
@fisherxu Thanks for your correction! Finished!

File: keadm/cmd/keadm/app/cmd/reset.go [hunk 7]
```diff
@@ -18,13 +18,15 @@ package cmd
...
+	"io"
 
 	"github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/common"
-	types "github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/common"
 	"github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/util"
+	phases "k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/reset"
+	types "github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/common"
+	utilruntime "k8s.io/kubernetes/cmd/kubeadm/app/util/runtime"
+	utilsexec "k8s.io/utils/exec"
 )
```
Comment (fisherxu) [inline keadm/cmd/keadm/app/cmd/reset.go:30]:
```
	"fmt"
	"io"

	"github.com/spf13/cobra"
	phases "k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/reset"
	utilruntime "k8s.io/kubernetes/cmd/kubeadm/app/util/runtime"
	utilsexec "k8s.io/utils/exec"

      "github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/util"
       types "github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/common"
```

File: keadm/cmd/keadm/app/cmd/reset.go [hunk 8]
```diff
@@ -103,6 +118,36 @@ func TearDownKubeEdge(isEdgeNode bool, kubeConfig string) error {
...
+	if isEdgeNode {
+		dirToClean = append(dirToClean, "/var/lib/dockershim", "/var/run/kubernetes", "/var/lib/cni")
+	}
+
+	// TODO: more mount directories?
+
+	for _, dir := range dirToClean {
+		_ = phases.CleanDir(dir)
+	}
+	return nil
```
Comment (fisherxu) [inline keadm/cmd/keadm/app/cmd/reset.go:148]: Please handle all errors in `RemoveContainers` and `cleanDirectories`

File: keadm/cmd/keadm/app/cmd/reset.go [hunk 9]
```diff
@@ -124,23 +124,11 @@ func RemoveContainers(isEdgeNode bool, execer utilsexec.Interface) error {
...
-	}
-	fmt.Println("[Test] Got containerRuntime", containerRuntime)
+	containerRuntime, _ := utilruntime.NewContainerRuntime(execer, criSocketPath)
 
-	containers, err := containerRuntime.ListKubeContainers()
-	if err != nil {
-		return err
-	}
-	fmt.Println("[Test] Got containers", containers)
+	containers, _ := containerRuntime.ListKubeContainers()
```
Comment (GsssC) [inline keadm/cmd/keadm/app/cmd/reset.go:131]: handle error and use `klog` to print out.

File: keadm/cmd/keadm/app/cmd/reset.go (hunk 7 above)
Comment (ESWZY) [reply keadm/cmd/keadm/app/cmd/reset.go:30]: Sorted.

File: keadm/cmd/keadm/app/cmd/reset.go (hunk 8 above)
Comment (ESWZY) [reply keadm/cmd/keadm/app/cmd/reset.go:148]: Added.

File: keadm/cmd/keadm/app/cmd/reset.go (hunk 9 above)
Comment (ESWZY) [reply keadm/cmd/keadm/app/cmd/reset.go:131]: Added.

**Comment by daixiang0** [comment]
@fisherxu @GsssC @ESWZY i think its root cause is relate to #2035 and #2011. 

`keadm reset` kills keadm with SIGKILL then `edgecore` can not shutdown gracefully, also `edgecore` does not cleanup containers which should not run.

So I think no need to need this pr.

**Comment by GsssC** [comment]
image
@daixiang0 But it seems that `gracefullshutdown` does not clean `k8s-related-containers`

**Comment by GsssC** [comment]
@daixiang0 even if `kubelet`
image

**Comment by daixiang0** [comment]
I think `kebeadm reset` would not cleanup all containers refert to [doc](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-reset/#reset-workflow).

**Comment by GsssC** [comment]
@daixiang0 yes, but according to source code, kubeadm would remove `k8s-related-containers` .
image

File: keadm/cmd/keadm/app/cmd/reset.go (hunk 6 above)
Comment (GsssC) [reply keadm/cmd/keadm/app/cmd/reset.go:28]: @ESWZY imports sort wrong

Comment (ESWZY) [reply keadm/cmd/keadm/app/cmd/reset.go:28]: @GsssC Sorted in commit d259691.

File: go.mod [hunk 10]
```diff
@@ -66,12 +66,13 @@ require (
 	google.golang.org/grpc v1.26.0
 	gopkg.in/square/go-jose.v2 v2.3.1 // indirect
 	gopkg.in/yaml.v2 v2.2.8
-	k8s.io/api v0.18.6
+	k8s.io/api v0.18.8
 	k8s.io/apiextensions-apiserver v0.18.6
-	k8s.io/apimachinery v0.18.6
+	k8s.io/apimachinery v0.18.8
```
Comment (fisherxu) [inline go.mod:71]: don't update the k8s version dependency

File: go.mod [hunk 11]
```diff
@@ -66,12 +66,13 @@ require (
...
 	k8s.io/apiserver v0.18.6
 	k8s.io/client-go v0.18.6
 	k8s.io/cloud-provider v0.18.6
+	k8s.io/cluster-bootstrap v0.18.8 // indirect
```
Comment (fisherxu) [inline go.mod:75]: any feature need upgrade to k8s 1.18.8？

**Comment by fisherxu** [comment]
Looks like the CI is still failed :)

**Comment by ESWZY** [comment]
> Looks like the CI is still failed :)

Hi! I have also noticed that. 

At beginning, I used the command `make verify` to check the project. It can pass the CI test, but updated k8s to v0.18.8.

This time, I reverted the changes about dependencies, but the CI failed.

Is there any method to deal with that? Thank you!

**Comment by fisherxu** [comment]
you should run `make verify`

**Comment by ESWZY** [comment]
> you should run `make verify`

I have run `make verify` in commit 21d2e2c.

And it changed the version of k8s.

**Comment by ESWZY** [comment]
/cc fisherxu 

Hello, it passed CI now. Could you give a review.

**Comment by fisherxu** [comment]
> /cc fisherxu
> 
> Hello, it passed CI now. Could you give a review.

Looks good, could you please squash the commits? @ESWZY

**Comment by ESWZY** [comment]
> > /cc fisherxu
> > Hello, it passed CI now. Could you give a review.
> 
> Looks good, could you please squash the commits? @ESWZY

Hello, squashed!

**Comment by fisherxu** [comment]
/lgtm
/approve
Thanks for the enhancement

**Pull request merged**